### PR TITLE
Show whitch exactly check action is bad encoded

### PR DESCRIPTION
**Pull request by SandakovMM**
Otherwise it's preatty hard to understand where the customer should look for a 'bad' encoded file.

### CODE REVIEW

File: pleskdistup/main.py [hunk 1]
```diff
@@ -101,14 +101,19 @@ def required_conditions_satisfied(upgrader: DistUpgrader, options: typing.Any, p
...
             writer.write("\r")
             for check in failed_checks:
-                writer.write(check)
-                log.err(check)
+                try:
+                    writer.write("Required pre-conversion condition {check.name!r} not met:\n\t{check.description}\n")
+                    log.err(check)
+                except (UnicodeEncodeError, UnicodeDecodeError) as ex:
+                    bad_encoded_check = check.name
+                    raise RuntimeError(f"Encoding error on writing problem description") from ex
```
Comment (kpushkaryov) [inline pleskdistup/main.py:116]: Pointless f-string.

File: pleskdistup/main.py [hunk 2]
```diff
@@ -101,14 +101,19 @@ def required_conditions_satisfied(upgrader: DistUpgrader, options: typing.Any, p
...
         with action.CheckFlow(checks) as check_flow, writers.StdoutWriter() as writer:
             writer.write("Doing preparation checks...\n")
             check_flow.validate_actions()
             failed_checks = check_flow.make_checks()
             writer.write("\r")
             for check in failed_checks:
-                writer.write(check)
-                log.err(check)
+                try:
+                    writer.write("Required pre-conversion condition {check.name!r} not met:\n\t{check.description}\n")
```
Comment (kpushkaryov) [inline pleskdistup/main.py:112]: Omitted f-string.

**Review by kpushkaryov — CHANGES_REQUESTED** [review]
Typo in the commit message: whitch -> which.

This output errors processing is getting messy. Could we instead use an alternative [error handler](https://docs.python.org/3/library/functions.html#open), like `replace`?

**Comment by SandakovMM** [comment]
> This output errors processing is getting messy. Could we instead use an alternative [error handler](https://docs.python.org/3/library/functions.html#open), like replace?

We could use 'replace' for pre-checks. But then we would be hiding the inconsistency in locale from the user.
I would like to clearly identify the problem.

**Comment by kpushkaryov** [comment]
> We could use 'replace' for pre-checks. But then we would be hiding the inconsistency in locale from the user. I would like to clearly identify the problem.

I mean, use `replace` or its analog for terminal and log output. Otherwise, we can't output the problematic message itself, even if it's 99% ASCII, thus reducing diagnostic potential.

**Comment by SandakovMM** [comment]
> I mean, use replace or its analog for terminal and log output

Yep, but unfortunately there is no errors argument for logging.FileHandler in python 3.6. So we can't simple change error handling for logs.

File: pleskdistup/common/src/writers.py [hunk 3]
```diff
@@ -29,6 +29,28 @@ def __exit__(self, *args):
...
+
+    def write(self, message: str) -> None:
+        self.out.write(message)
+        self.out.flush()
+
+        recoded_message = codecs.encode(message, self.out.encoding, errors='replace').decode(self.out.encoding)
+        if recoded_message != message:
+            self.encoding_errors_met = True
+
+    def is_there_encoding_errors(self) -> bool:
```
Comment (kpushkaryov) [inline pleskdistup/common/src/writers.py:47]: `is_there_encoding_errors` is weird, I suggest `has_encoding_errors`.

File: pleskdistup/common/src/log.py [hunk 4]
```diff
@@ -9,10 +10,20 @@ class logger():
 
     is_streams_enabled = False
     streams_logger = logging.getLogger("distupgrade_streams")
+    encoding: str = locale.getpreferredencoding()
+
+    @staticmethod
+    def _re_decode_message(message: str) -> str:
+        return str(message).encode(logger.encoding, errors='backslashreplace').decode(logger.encoding, errors='backslashreplace')
```
Comment (kpushkaryov) [inline pleskdistup/common/src/log.py:17]: According to annotation, `message` should be also `str`, why `str()` then?

File: pleskdistup/main.py [hunk 5]
```diff
@@ -101,19 +101,19 @@ def required_conditions_satisfied(upgrader: DistUpgrader, options: typing.Any, p
...
-                    log.err(check)
-                except (UnicodeEncodeError, UnicodeDecodeError) as ex:
-                    bad_encoded_check = check.name
-                    raise RuntimeError(f"Encoding error on writing problem description") from ex
+                writer.write(f"Required pre-conversion condition {check.name!r} not met:\n\t{check.description}\n")
+                log.err(check)
+
+            if writer.is_there_encoding_errors():
+                printerr(messages.ENCODING_INCONSISTENCY_ERROR_MESSAGE)
+                printerr("The encoding problem occurred during one of the previous check. The badly encoded symbol was replaced with a backslash Unicode representation")
```
Comment (kpushkaryov) [inline pleskdistup/main.py:115]: check -> checks

**Review by kpushkaryov — CHANGES_REQUESTED** [review]

File: pleskdistup/common/src/log.py (hunk 4 above)
Comment (SandakovMM) [reply pleskdistup/common/src/log.py:17]: We sometimes send objects into log functions, like in this [example] (https://github.com/plesk/dist-upgrader/blob/main/pleskdistup/main.py#L111) , where a failed pre-check is logged to the error log.
Likely we could just change type annotation to "something convertible to string"

Comment (SandakovMM) [reply pleskdistup/common/src/log.py]: Ah, wait a second, the place was actually broken by this change, we just need to fix it

**Review by kpushkaryov — DISMISSED** [review]
Please rebase before merge. Don't forget to fix the typo in the commit message.

**Review by kpushkaryov — APPROVED** [review]

**Pull request merged**